explorer/interpreter/type_checker.cpp

```diff
@@ -2335,10 +2336,28 @@ auto TypeChecker::MatchImpl(const InterfaceType& iface,
       *trace_stream_ << "matched with " << *impl.type << " as "
                      << *impl.interface << "\n\n";
     }
-    CARBON_ASSIGN_OR_RETURN(
-        const auto* subst_witness,
-        SubstituteCast<Witness>(*bindings_or_error, impl.witness));
-    return {subst_witness};
+
+    auto subst_witness_or_error =
+        SubstituteCast<Witness>(*bindings_or_error, impl.witness);
+
+    if (!subst_witness_or_error.ok()) {
+      std::string error;
+      llvm::raw_string_ostream stream(error);
+
+      stream << subst_witness_or_error.error().message() << "\n"
+             << "NOTE: " << source_loc.ToString()
+             << ": in instantiation of impl ";
+      impl.type->Print(stream);
+      stream << " as " << GetName(impl.interface->declaration()) << " with ";
+      impl.type->Print(stream);
+      stream << " = ";
+      impl_type->Print(stream);
```

Comment:
I am not really sure about the calls to `Print` for `impl.type` and `impl_type` as they seem more suitable for `trace_stream_` and not for error reporting. Is there any other way I can get the string representation of a type?

Reply:
Thanks for the suggestion, the code looks definitely cleaner.